language-plutus-core/stdlib/Language/PlutusCore/StdLib/Meta/Data/Tuple.hs

```diff
@@ -3,104 +3,163 @@
+--         -- all R
+--         TyForall () resultType (Type ()) $
+--         -- (T_1 -> .. -> T_n -> r) -> r
+--         TyFun () caseType (TyVar () resultType)
+
+-- -- | Given an arity @n@, create the constructor for n-tuples.
+-- --
+-- -- @
+-- --     /\(T_1 :: *) .. (T_n :: *) .
+-- --         \(arg_1 : T_1) .. (arg_n : T_n) .
+-- --             /\(R :: *).
+-- --                 \(case : T_1 -> .. -> T_n -> R) -> case arg_1 .. arg_n
+-- -- @
+-- getBuiltinTupleConstructor :: MonadQuote m => Int -> m (Term TyName Name ())
+-- getBuiltinTupleConstructor arity = do
+--     tyVars <- for [0..(arity-1)] $ \i -> do
+--         tn <- liftQuote $ freshTyName () $ strToBs $ "t_" ++ show i
+--         pure $ TyVarDecl () tn $ Type ()
+
+--     resultType <- liftQuote $ freshTyName () "r"
+
+--     args <- for [0..(arity -1)] $ \i -> do
+--         n <- liftQuote $ freshName () $ strToBs $ "arg_" ++ show i
+--         pure $ VarDecl () n $ mkTyVar () $ tyVars !! i
+
+--     caseArg <- liftQuote $ freshName () "case"
+--     let caseTy = mkIterTyFun () (fmap (mkTyVar ()) tyVars) (TyVar () resultType)
+--     pure $
+--         -- /\T_1 .. T_n
+--         mkIterTyAbs () tyVars $
+--         -- \arg_1 .. arg_n
+--         mkIterLamAbs () args $
+--         -- /\R
+--         TyAbs () resultType (Type ()) $
+--         -- \case
+--         LamAbs () caseArg caseTy $
+--         -- case arg_1 .. arg_n
+--         mkIterApp () (Var () caseArg) $ fmap (mkVar ()) args
+
+-- -- | Given an arity @n@ and an index @i@, create a function for accessing the i'th component of a n-tuple.
+-- --
+-- -- @
+-- --     /\(T_1 :: *) .. (T_n :: *) .
+-- --         \(tuple : all (R :: *) . (T_1 -> .. -> T_n -> R) -> R)) .
+-- --             tuple {T_i} (\(arg_1 : T_1) .. (arg_n : T_n) . arg_n)
+-- -- @
+-- getBuiltinTupleAccessor :: MonadQuote m => Int -> Int -> m (Term TyName Name ())
+-- getBuiltinTupleAccessor arity index = rename =<< do
+--     tyVars <- for [0..(arity-1)] $ \i -> do
+--         tn <- liftQuote $ freshTyName () $ strToBs $ "t_" ++ show i
+--         pure $ TyVarDecl () tn $ Type ()
+
+--     tupleTy <- do
+--         genericTuple <- getBuiltinTuple arity
+--         pure $ mkIterTyApp () genericTuple (fmap (mkTyVar ()) tyVars)
+--     let selectedTy = mkTyVar () $ tyVars !! index
+
+--     args <- for [0..(arity -1)] $ \i -> do
+--         n <- liftQuote $ freshName () $ strToBs $ "arg_" ++ show i
+--         pure $ VarDecl () n $ mkTyVar () $ tyVars !! i
+--     let selectedArg = mkVar () $ args !! index
+
+--     tupleArg <- liftQuote $ freshName () "tuple"
+--     pure $
+--         -- /\T_1 .. T_n
+--         mkIterTyAbs () tyVars $
+--         -- \tuple :: (tupleN T_1 .. T_n)
+--         LamAbs () tupleArg tupleTy $
+--         -- tuple {T_i}
+--         Apply () (TyInst () (Var () tupleArg) selectedTy) $
+--         -- \arg_1 .. arg_n . arg_i
+--         mkIterLamAbs () args selectedArg
+
+data Tuple ann = Tuple
```

Comment:
Well, this is the old discussion about how much should be meta and how much should be in PLC. The previous approach of this module was to have the minimum meta - only the arity, which would have been quite annoying to do in PLC. This puts much more into meta. Not sure which is better tbh.

Reply:
> The previous approach of this module was to have the minimum meta - only the arity

Library-wise, yes. But when we actually need to use that library, we have to keep the necessary meta-information in order to instantiate type abstractions.

So we have that `getBuiltinMutualFixOf` function. It returns the fixed point of a family of mutually recursive functions as a Scott-encoded tuple. But if we kept only the term, it would mean that we can't actually do much with the tuple. For example, you can't really take the ith element of a tuple without knowing the types of its elements. Of course, you can write the generic accessor function like the one we had before, but this is just asking the user to keep track of types and provide them later. Those are details that we can remember and free the user from them: `getBuiltinMutualFixOf` returns an encapsulated tuple, elements of which can be accessed without the need to explicitly pass types around.

Previously it was

```
    user
  /      \
tuple  element types
  \      /
   result
```

now it's

```
    user
     |
   tuple
     |
   result
```

But this is "theoretically". Practically, there is not much difference. The user still has to supply names to `bindTuple` (tuples are anonymous and it doesn't make sense to hardcode names for elements) where he previously need to supply typed names. I anyway think this is a more fine-grained solution, because previously we were forcing the user to supply types twice: when constructing a tuple and when accessing it, while now we require to supply types only once. And the new solution allows to cut some corners and generate less type abstractions/instantiations and is also much more convenient when you write Plutus Core by hand.

Reply:
It would be good to have functions around for creating the tuple types, constructors etc. themselves, since that can be handy if you actually want to pass that as an argument or something.